File: server/models/media_share.go
package models

import "time"

// MediaShare est un lien de partage public tel que son créateur le voit
// (GET /api/shares, POST /api/shares). Voir ADR-0037.
type MediaShare struct {
	ID        int    `json:"id"`
	MediaID   int    `json:"media_id"`
	MediaType string `json:"media_type"`
	// Title est le film, ou la série pour un épisode ; Subtitle porte alors
	// « S01E02 · Titre de l'épisode ».
	Title       string     `json:"title"`
	Subtitle    string     `json:"subtitle,omitempty"`
	PosterURL   string     `json:"poster_url,omitempty"`
	HasPassword bool       `json:"has_password"`
	SingleUse   bool       `json:"single_use"`
	ExpiresAt   *time.Time `json:"expires_at,omitempty"`
	// Claimed : un navigateur a réservé ce lien à usage unique.
	Claimed    bool       `json:"claimed"`
	ConsumedAt *time.Time `json:"consumed_at,omitempty"`
	Views      int        `json:"views"`
	// Status vaut "active", "expired" ou "watched".
	Status    string    `json:"status"`
	CreatedAt time.Time `json:"created_at"`
	// Code n'est rendu qu'à la création : le serveur n'en garde que
	// l'empreinte, il ne pourra plus le redonner.
	Code string `json:"code,omitempty"`
}

// SharedMedia est ce qu'un visiteur sans compte apprend d'un lien
// (POST /api/shared/info). Tant qu'un mot de passe est exigé, le média reste
// anonyme : Title et les suivants sont vides.
type SharedMedia struct {
	NeedsPassword bool       `json:"needs_password"`
	SingleUse     bool       `json:"single_use"`
	ExpiresAt     *time.Time `json:"expires_at,omitempty"`
	MediaType     string     `json:"media_type,omitempty"`
	Title         string     `json:"title,omitempty"`
	Subtitle      string     `json:"subtitle,omitempty"`
	PosterURL     string     `json:"poster_url,omitempty"`
	Duration      int        `json:"duration,omitempty"`
}

// SharedMediaAccess est la réponse à l'ouverture d'un lien
// (POST /api/shared/open) : un ticket de lecture du média, et le jeton du
// navigateur à présenter pour le renouveler.
type SharedMediaAccess struct {
	Media             SharedMedia `json:"media"`
	MediaID           int         `json:"media_id"`
	Ticket            string      `json:"ticket"`
	ExpiresAt         time.Time   `json:"expires_at"`
	RenewAfterSeconds int         `json:"renew_after_seconds"`
	Viewer            string      `json:"viewer"`
}

// SharedMediaRenewal répond à POST /api/shared/renew.
type SharedMediaRenewal struct {
	ExpiresAt time.Time `json:"expires_at"`
}

// SharedMediaProgress répond à POST /api/shared/progress. Consumed dit que le
// lien, à usage unique, est désormais détruit.
type SharedMediaProgress struct {
	Consumed bool `json:"consumed"`
}

File: app/lib/models/media_share.dart
/// Un lien de partage public vers un film ou un épisode (ADR-0037), tel que
/// son créateur le voit. Miroir de `models.MediaShare` côté serveur.
class MediaShare {
  const MediaShare({
    required this.id,
    required this.mediaId,
    required this.title,
    required this.status,
    this.mediaType = '',
    this.subtitle = '',
    this.posterUrl,
    this.hasPassword = false,
    this.singleUse = false,
    this.expiresAt,
    this.claimed = false,
    this.consumedAt,
    this.views = 0,
    this.createdAt,
    this.code,
  });

  final int id;
  final int mediaId;
  final String mediaType;

  /// Le film, ou la série pour un épisode ; [subtitle] porte alors
  /// « S01E02 · Titre de l'épisode ».
  final String title;
  final String subtitle;
  final String? posterUrl;
  final bool hasPassword;
  final bool singleUse;

  /// Nul pour un lien sans échéance.
  final DateTime? expiresAt;

  /// Un navigateur a réservé ce lien à usage unique.
  final bool claimed;
  final DateTime? consumedAt;
  final int views;

  /// `active`, `expired` ou `watched`, décidé par le serveur.
  final String status;
  final DateTime? createdAt;

  /// Le code du lien. Le serveur ne le rend qu'à la création : il n'en garde
  /// que l'empreinte, et ne pourra plus le redonner.
  final String? code;

  bool get isActive => status == 'active';

  factory MediaShare.fromJson(Map<String, dynamic> json) {
    DateTime? parse(dynamic raw) => raw is String && raw.isNotEmpty
        ? DateTime.tryParse(raw)?.toLocal()
        : null;
    final code = json['code'] as String?;
    return MediaShare(
      id: json['id'] as int? ?? 0,
      mediaId: json['media_id'] as int? ?? 0,
      mediaType: json['media_type'] as String? ?? '',
      title: json['title'] as String? ?? '',
      subtitle: json['subtitle'] as String? ?? '',
      posterUrl: json['poster_url'] as String?,
      hasPassword: json['has_password'] == true,
      singleUse: json['single_use'] == true,
      expiresAt: parse(json['expires_at']),
      claimed: json['claimed'] == true,
      consumedAt: parse(json['consumed_at']),
      views: json['views'] as int? ?? 0,
      status: json['status'] as String? ?? 'active',
      createdAt: parse(json['created_at']),
      code: code == null || code.isEmpty ? null : code,
    );
  }
}

/// Les durées de validité qu'un lien peut avoir, dans l'ordre où elles sont
/// proposées. La liste est fermée côté serveur (`sharelinks.Lifetimes`).
enum ShareLifetime {
  day(24, '24 heures'),
  week(24 * 7, '7 jours'),
  month(24 * 30, '30 jours'),
  forever(0, 'Sans limite');

  const ShareLifetime(this.hours, this.label);

  /// La valeur envoyée au serveur ; 0 veut dire « sans échéance ».
  final int hours;
  final String label;
}

/// Ce qu'un visiteur sans compte apprend d'un lien (POST /api/shared/info et
/// /open). Tant qu'un mot de passe est exigé, [title] et les suivants sont
/// vides : le serveur ne dit pas quel média le lien ouvre.
class SharedMediaInfo {
  const SharedMediaInfo({
    this.needsPassword = false,
    this.singleUse = false,
    this.expiresAt,
    this.mediaType = '',
    this.title = '',
    this.subtitle = '',
    this.posterUrl,
    this.duration = 0,
  });

  final bool needsPassword;
  final bool singleUse;
  final DateTime? expiresAt;
  final String mediaType;
  final String title;
  final String subtitle;
  final String? posterUrl;
  final int duration;

  factory SharedMediaInfo.fromJson(Map<String, dynamic> json) {
    final expires = json['expires_at'];
    final poster = json['poster_url'] as String?;
    return SharedMediaInfo(
      needsPassword: json['needs_password'] == true,
      singleUse: json['single_use'] == true,
      expiresAt:
          expires is String ? DateTime.tryParse(expires)?.toLocal() : null,
      mediaType: json['media_type'] as String? ?? '',
      title: json['title'] as String? ?? '',
      subtitle: json['subtitle'] as String? ?? '',
      posterUrl: poster == null || poster.isEmpty ? null : poster,
      duration: json['duration'] as int? ?? 0,
    );
  }
}

/// Un refus du serveur sur un lien : [message] est la phrase à montrer telle
/// quelle au visiteur, [status] le code HTTP (404 inconnu, 410 expiré ou vu,
/// 409 ouvert ailleurs, 401 mauvais mot de passe, 0 serveur injoignable).
class SharedLinkException implements Exception {
  const SharedLinkException(this.status, this.message);

  final int status;
  final String message;

  @override
  String toString() => 'SharedLinkException($status, $message)';
}
